### PR TITLE
Agent implementation

### PR DESCRIPTION
**Pull request by rsauex**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.example</groupId>
```
Comment (fixedorgo) [inline pom.xml]: We should change artifact data also according to project name and package.

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <artifactId>maven-assembly-plugin</artifactId>
+                <version>2.2</version>
+                <configuration>
+                    <descriptorRefs>
+                        <descriptorRef>jar-with-dependencies</descriptorRef>
```
Comment (fixedorgo) [inline pom.xml:36]: Maybe rename it to `uberjar`? 😃

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+public final class AddRequireToClinitClassVisitor extends ClassVisitor {
+    private final ClassLoader loader;
+
+    private String className;
+    private int version;
+    private boolean skip;
+
+    private boolean visitedStaticBlock = false;
+
+    AddRequireToClinitClassVisitor(int api, ClassVisitor visitor, ClassLoader loader) {
```
Comment (fixedorgo) [inline src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: I think this constructor should be `public`. Or not?

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+        super.visit(version, access, name, signature, superName, interfaces);
+        this.className = name;
+        this.version = version;
+        String scriptBase = className.split("\\$")[0];
+        skip = getResource(loader, scriptBase + "__init.class") == null
+                && getResource(loader, scriptBase + ".clj") == null
+                && getResource(loader, scriptBase + ".cljc") == null;
+//        if (!skip) {
```
Comment (fixedorgo) [inline src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: I think we can freely remove it or uncomment.

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import org.objectweb.asm.Label;
+import org.objectweb.asm.MethodVisitor;
+
+import static org.objectweb.asm.Opcodes.*;
+
+public final class AddRequireToClinitMethodVisitor extends MethodVisitor {
+    private final String className;
+    private final int version;
+
+    AddRequireToClinitMethodVisitor(int api, MethodVisitor methodVisitor, String className, int version) {
```
Comment (fixedorgo) [inline src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java]: Same here.

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        super.visitMethodInsn(INVOKEINTERFACE, "clojure/lang/IFn", "invoke", "(Ljava/lang/Object;)Ljava/lang/Object;", true);
+        super.visitInsn(POP);
+
+        super.visitJumpInsn(GOTO, end);
+
+        super.visitLabel(handler);
+        if (stackFramesEnabled()) {
+            super.visitFrame(F_FULL, 0, new Object[]{}, 1, new Object[]{"java/lang/Exception"});
+        }
+//        super.visitMethodInsn(INVOKEVIRTUAL, "java/lang/Throwable", "printStackTrace", "()V", false);
```
Comment (fixedorgo) [inline src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java]: Should we keep this?

**Review by fixedorgo — CHANGES_REQUESTED** [review]

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 4 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: How about checking an environment variable (e.g. `DEBUG_EMR_JAVA_AGENT`) so the message is displayed only when we what it to?

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 3 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: This class should not be used by anyone except for the Agent itself. I'd rather make the class package-private?

File: pom.xml (hunk 2 above)
Comment (rsauex) [reply pom.xml:36]: There is an XML file in the maven-assembly-plugin which defines `jar-with-dependencies`. I'd rather not copy it into the repo just for the sake of a different name.

File: pom.xml (hunk 1 above)
Comment (rsauex) [reply pom.xml]: Done

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 3 above)
Comment (fixedorgo) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: Well it's a package private already. But usually package private constructors looks weird especially when this is your library and any project person can change the access. There are also some special cases with inheritance.

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 4 above)
Comment (fixedorgo) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: Sure. Any useful logging better than a commented block of code. 😃

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 3 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: It's public now

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java (hunk 5 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java]: It's public now

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java (hunk 4 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitClassVisitor.java]: Done

File: src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java (hunk 6 above)
Comment (rsauex) [reply src/main/java/com/apptopia/emr/javaagent/AddRequireToClinitMethodVisitor.java]: Using `DEBUG_EMR_JAVA_AGENT`

**Review by fixedorgo — APPROVED** [review]
Looks good! 👍

**Pull request merged**